mathoxide-lib/src/shape_utils.rs

```diff
@@ -0,0 +1,58 @@
+pub enum ShapeDim {
+    Known(usize),
+    Dyn,
+}
+
+impl From<isize> for ShapeDim {
+    fn from(val: isize) -> Self {
+        if val >= 0 {
+            ShapeDim::Known(val as usize)
+        } else {
+            ShapeDim::Dyn
+        }
+    }
+}
+
+impl From<usize> for ShapeDim {
+    fn from(val: usize) -> Self {
+        ShapeDim::Known(val)
+    }
+}
+
+fn parse<T, ListType>(shape: ListType) -> Vec<ShapeDim>
```

Comment:
Since `parse` is a generic function in the standard library, this name might be confusing. Perhaps a name like `normalize_shape` might be more expressive?